detect mocked module on saving pass

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69793

Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

PaliC committed:
detect mocked module on saving pass

[ghstack-poisoned]

suo commented on torch/package/package_exporter.py:
```diff
@@ -39,7 +39,6 @@
 
 ActionHook = Callable[["PackageExporter", str], None]
 
-
```
nit: try not to include unrelated changes like whitespace and formatting in your PR to keep the project history clean.

PaliC replied:
ah gotcha, so just lint right before merging then

suo commented on torch/package/package_exporter.py:
```diff
@@ -550,7 +549,10 @@ def _intern_module(
+        self, package: str,
+        resource: str, obj: Any,
+        dependencies: bool = True,
+        save_mocked_and_used: bool = False
```
I think we don't need to make this an option for users. While this change is technically BC-incompatible, any package that trips this behavior would not have been importable anyway, so it's probably fine.

PaliC replied:
Yeah it was primarily for the unit test, I couldn't think of a better way to do the bc test. I'll just delete it then.

suo reviewed (changes requested):
Looks pretty good overall, only a small comment about not presenting it as an option in the API. We should make sure internal tests pass on this as well

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

[ghstack-poisoned]

suo commented:
Please import these changes internally so that we can run fbcode unit tests against them

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

suo reviewed (approved):
lgtm, but make sure we don't break anything internal :)

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

PaliC committed:
Update on "detect mocked module on saving pass"


Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).
